File: ios/NeuralComposeClient/Sources/EEGScreen.swift
// The one working M4 slice: EEG stream status rendered PURELY from the core's
// Presentation + ChannelSnapshot. Labels/banners come from the core's English
// formatters (semantic state is available for localized shells later).
//
// Plus the Muse golden-capture gate controls: endpoint, record, and the
// published recordings with per-recording replay verification.

import NeuralComposeCore
import SwiftUI

struct EEGScreen: View {
    @StateObject private var model = EEGStreamModel()

    private var toneColor: Color {
        switch model.presentation.tone {
        case .ok: return .green
        case .stale, .connecting: return .orange
        case .down: return .red
        }
    }

    var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: 16) {
                header
                Text("\(model.snapshot.received) samples received")
                    .font(.caption).foregroundStyle(.secondary)

                if let banner = formatBannerEn(p: model.presentation) {
                    Text(banner)
                        .font(.callout.weight(.semibold))
                        .frame(maxWidth: .infinity)
                        .padding(10)
                        .background(
                            RoundedRectangle(cornerRadius: 8)
                                .strokeBorder(toneColor)
                                .background(toneColor.opacity(0.15))
                        )
                }

                endpointSection
                captureSection
                channels
                recordingsSection
            }
            .padding()
        }
        .onAppear { model.start() }
        .onDisappear { model.stop() }
    }

    private var header: some View {
        HStack {
            Text("EEG Stream").font(.largeTitle.bold())
            Spacer()
            HStack(spacing: 6) {
                Circle().fill(toneColor).frame(width: 8, height: 8)
                Text(formatLabelEn(p: model.presentation))
                    .font(.caption.bold()).monospaced()
            }
            .padding(.horizontal, 10).padding(.vertical, 4)
            .background(Capsule().strokeBorder(toneColor))
        }
    }

    // MARK: - endpoint

    private var endpointSection: some View {
        VStack(alignment: .leading, spacing: 8) {
            Text("Bridge endpoint").font(.headline)
            TextField("ws://host:port/api/eeg/stream", text: $model.endpointText)
                .textFieldStyle(.roundedBorder)
                .font(.callout.monospaced())
                .keyboardType(.URL)
                .autocorrectionDisabled()
                .textInputAutocapitalization(.never)
                .submitLabel(.go)
                .onSubmit { model.applyEndpoint() }
            HStack {
                Button("Connect") { model.applyEndpoint() }
                    .buttonStyle(.borderedProminent)
                    .disabled(model.isRecording)
                Spacer()
                Text(model.connectedEndpoint)
                    .font(.caption2.monospaced()).foregroundStyle(.secondary)
                    .lineLimit(1).truncationMode(.head)
            }
            if let error = model.endpointError {
                Text(error).font(.caption).foregroundStyle(.red)
            }
            Text("On a physical phone 127.0.0.1 is the phone itself — use the bridge's LAN address.")
                .font(.caption2).foregroundStyle(.tertiary)
        }
        .padding(12)
        .background(RoundedRectangle(cornerRadius: 10).fill(.quaternary.opacity(0.25)))
    }

    // MARK: - capture

    private var captureSection: some View {
        VStack(alignment: .leading, spacing: 8) {
            HStack {
                Text("Golden capture").font(.headline)
                Spacer()
                Button(model.isRecording ? "Stop" : "Start recording") {
                    if model.isRecording { model.stopRecording() } else { model.startRecording() }
                }
                .buttonStyle(.borderedProminent)
                .tint(model.isRecording ? .red : .accentColor)
            }

            if let stats = model.captureStats {
                VStack(alignment: .leading, spacing: 2) {
                    Text(stats.recordingId).font(.caption.monospaced().bold())
                    Text(
                        "\(stats.messagesReceived) messages · \(stats.acceptedSampleCount) samples · "
                            + "\(model.elapsedMs / 1000)s"
                    )
                    .font(.caption.monospaced())
                    if let failure = stats.writeFailure {
                        Text("WRITE FAILED — will discard on stop: \(failure)")
                            .font(.caption.bold()).foregroundStyle(.red)
                    }
                    Text("Stream stays connected while recording, even off this tab.")
                        .font(.caption2).foregroundStyle(.tertiary)
                }
            }

            if let message = model.captureMessage {
                Text(message).font(.caption).foregroundStyle(.secondary)
            }
        }
        .padding(12)
        .background(RoundedRectangle(cornerRadius: 10).fill(.quaternary.opacity(0.25)))
    }

    // MARK: - channels

    private var channels: some View {
        ForEach(
            Array(zip(["TP9", "AF7", "AF8", "TP10"], model.snapshot.channels)),
            id: \.0
        ) { name, values in
            VStack(alignment: .leading, spacing: 4) {
                Text(name).font(.caption.bold())
                Sparkline(values: Array(values.suffix(256)))
                    .stroke(name.hasPrefix("AF") ? Color.green : Color.blue, lineWidth: 1)
                    .frame(height: 64)
                    .background(
                        RoundedRectangle(cornerRadius: 8).fill(.quaternary.opacity(0.3)))
            }
        }
    }

    // MARK: - recordings

    private var recordingsSection: some View {
        VStack(alignment: .leading, spacing: 10) {
            HStack {
                Text("Recordings").font(.headline)
                Spacer()
                Button("Refresh") { model.reloadRecordings() }.font(.caption)
            }

            if model.listing.recordings.isEmpty {
                Text("No published recordings on this device.")
                    .font(.caption).foregroundStyle(.secondary)
            }

            ForEach(model.listing.recordings) { recording in
                RecordingRow(
                    recording: recording,
                    verdict: model.verdicts[recording.id],
                    verify: { model.verify(recording) },
                    delete: { model.delete(recording) })
            }

            if !model.listing.unreadable.isEmpty {
                Text("Unreadable manifests: \(model.listing.unreadable.joined(separator: ", "))")
                    .font(.caption).foregroundStyle(.red)
            }
            if !model.listing.partials.isEmpty {
                Text(
                    "\(model.listing.partials.count) incomplete .partial file(s) from an interrupted run — not recordings."
                )
                .font(.caption2).foregroundStyle(.orange)
            }
        }
        .padding(12)
        .background(RoundedRectangle(cornerRadius: 10).fill(.quaternary.opacity(0.25)))
    }
}

private struct RecordingRow: View {
    let recording: CapturedRecording
    let verdict: String?
    let verify: () -> Void
    let delete: () -> Void

    var body: some View {
        VStack(alignment: .leading, spacing: 4) {
            Text(recording.id).font(.caption.monospaced().bold())
            Text(
                "\(recording.manifest.durationMs) ms · \(recording.manifest.acceptedSampleCount) samples · "
                    + "\(recording.manifest.payloadByteSize) B · \(recording.manifest.messagesReceived) msgs"
            )
            .font(.caption2.monospaced()).foregroundStyle(.secondary)
            Text(
                "\(recording.manifest.platform) \(recording.manifest.osVersion) · "
                    + "\(CaptureStore.string(for: recording.manifest.bridgeLocality)) · "
                    + "sha \(recording.manifest.payloadSha256Hex.prefix(12))"
            )
            .font(.caption2.monospaced()).foregroundStyle(.tertiary)

            if let note = recording.integrityNote {
                Text(note).font(.caption2.bold()).foregroundStyle(.red)
            }
            if let verdict {
                Text(verdict)
                    .font(.caption2)
                    .foregroundStyle(verdict.hasPrefix("Verified") ? .green : .red)
            }

            HStack {
                Button("Verify", action: verify).buttonStyle(.bordered).font(.caption)
                Button("Delete", role: .destructive, action: delete)
                    .buttonStyle(.bordered).font(.caption)
            }
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(10)
        .background(RoundedRectangle(cornerRadius: 8).fill(.quaternary.opacity(0.3)))
    }
}

struct Sparkline: Shape {
    let values: [Double]

    func path(in rect: CGRect) -> Path {
        var path = Path()
        guard values.count > 1 else { return path }
        let minV = values.min() ?? 0
        let maxV = values.max() ?? 1
        let span = max(maxV - minV, 1e-9)
        let stepX = rect.width / CGFloat(values.count - 1)
        for (i, v) in values.enumerated() {
            let x = CGFloat(i) * stepX
            let y = rect.height * (1 - CGFloat((v - minV) / span))
            if i == 0 { path.move(to: CGPoint(x: x, y: y)) } else {
                path.addLine(to: CGPoint(x: x, y: y))
            }
        }
        return path
    }
}

File: android/app/src/main/kotlin/org/neuralcompose/client/EEGScreen.kt
// The working M5-B slice (port of ios/.../EEGScreen.swift): rendered PURELY
// from the core's Presentation + ChannelSnapshot; labels/banners come from
// the core's English formatters.

package org.neuralcompose.client

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.lifecycle.viewmodel.compose.viewModel
import uniffi.neuralcompose_mobile_core.StreamTone
import uniffi.neuralcompose_mobile_core.formatBannerEn
import uniffi.neuralcompose_mobile_core.formatLabelEn

private val CHANNEL_NAMES = listOf("TP9", "AF7", "AF8", "TP10")

private fun toneColor(tone: StreamTone): Color = when (tone) {
    StreamTone.OK -> Color(0xFF2E9E4F)
    StreamTone.STALE, StreamTone.CONNECTING -> Color(0xFFE08A00)
    StreamTone.DOWN -> Color(0xFFD0342C)
}

@Composable
fun EEGScreen(model: EEGStreamViewModel = viewModel()) {
    LaunchedEffect(Unit) { model.start() }
    val ui by model.state.collectAsState()
    val tone = toneColor(ui.presentation.tone)

    Column(
        modifier = Modifier
            .fillMaxSize()
            .verticalScroll(rememberScrollState())
            .padding(16.dp),
        verticalArrangement = Arrangement.spacedBy(12.dp),
    ) {
        Row(verticalAlignment = Alignment.CenterVertically) {
            Text("EEG Stream", fontSize = 28.sp, fontWeight = FontWeight.Bold)
            Spacer(Modifier.weight(1f))
            Row(
                verticalAlignment = Alignment.CenterVertically,
                modifier = Modifier
                    .border(1.dp, tone, RoundedCornerShape(50))
                    .padding(horizontal = 10.dp, vertical = 4.dp),
            ) {
                Box(
                    Modifier
                        .size(8.dp)
                        .background(tone, CircleShape),
                )
                Spacer(Modifier.size(6.dp))
                Text(
                    formatLabelEn(ui.presentation),
                    fontSize = 12.sp,
                    fontWeight = FontWeight.Bold,
                    fontFamily = FontFamily.Monospace,
                )
            }
        }

        Text(
            "${ui.snapshot.received} samples received",
            fontSize = 13.sp,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )

        formatBannerEn(ui.presentation)?.let { banner ->
            Text(
                banner,
                fontSize = 15.sp,
                fontWeight = FontWeight.SemiBold,
                modifier = Modifier
                    .fillMaxWidth()
                    .border(1.dp, tone, RoundedCornerShape(8.dp))
                    .background(tone.copy(alpha = 0.12f), RoundedCornerShape(8.dp))
                    .padding(10.dp),
            )
        }

        ui.snapshot.channels.forEachIndexed { i, values ->
            Column(verticalArrangement = Arrangement.spacedBy(4.dp)) {
                Text(CHANNEL_NAMES[i], fontSize = 12.sp, fontWeight = FontWeight.Bold)
                Sparkline(
                    values = values.takeLast(256),
                    color = if (CHANNEL_NAMES[i].startsWith("AF")) Color(0xFF2E9E4F)
                    else Color(0xFF2B6FD9),
                )
            }
        }
    }
}

@Composable
private fun Sparkline(values: List<Double>, color: Color) {
    Canvas(
        modifier = Modifier
            .fillMaxWidth()
            .height(64.dp)
            .background(Color(0x14808080), RoundedCornerShape(8.dp)),
    ) {
        if (values.size < 2) return@Canvas
        val minV = values.min()
        val maxV = values.max()
        val span = (maxV - minV).coerceAtLeast(1e-9)
        val stepX = size.width / (values.size - 1)
        val path = Path()
        values.forEachIndexed { i, v ->
            val x = i * stepX
            val y = size.height * (1f - ((v - minV) / span).toFloat())
            if (i == 0) path.moveTo(x, y) else path.lineTo(x, y)
        }
        drawPath(path, color, style = Stroke(width = 2f))
    }
}
